### PR TITLE
Add API endpoints to update and fetch contact information

### PR DESCRIPTION
**Pull request by nima-karimi**
### Changes proposed in this Pull Request:

Part of https://github.com/woocommerce/google-listings-and-ads/issues/863. 

This is pointed to the `feature/contact-info` branch (so that we don't merge unfinished features to trunk) but that branch doesn't seem to contain the latest `trunk` changes so most of the changes here seem new but they're mainly based on Justin's PR https://github.com/woocommerce/google-listings-and-ads/pull/896.

The previously added API endpoints for updating and fetching the phone number are now modified to also handle updating the business address.

The endpoint name is changed from `phone-number` to `contact-information` but it still only accepts one `phone_number` input. However, now it also updates the address in Merchant Center to the store's address set in WooCommerce settings.

The `GET` endpoint now also returns three extra parameters in its response:
- `mc_address`: The address stored in Merchant Center
- `wc_address`: The address stored in WooCommerce Settings
- `is_mc_address_different`: A boolean indicating whether the two addresses above are different

This also sets the country for the Merchant Center account but given that the MC accounts created by GLA are all sub-accounts and [Google does not allow changing a sub-account's country](https://developers.google.com/shopping-content/reference/rest/v2.1/accounts#AccountAddress) we might need to modify this once the Google API behavior is changed. Google API currently doesn't enforce this and allows modifying the country (even though it's not possible to do so via MC interface).

Merchant Center API only accepts one address line and to adapt the WooCommerce address to it, the two addresses entered in WooCommerce settings are merged into one string separated with a new line character (`\n`). MC interface detects the addresses correctly and displays each in their own fields.

Note that address components are returned in the API response as they appear on Google API. This means that their naming might be different from WooCommerce. For example, city is called `locality`, state is called `region`, etc. This is just for the sake of simplicity and can be modified if we want to keep things similar to WC core.

### Detailed test instructions:

1. Connect a Merchant Center account and set a physical address for your store in WooCommerce settings.
2. Make a request to the `GET /wc/gla/mc/contact-information` endpoint
    - Confirm that no phone number or address is returned if you don't have anything in MC
    - if you already have a phone number and address set in MC make sure that it is returned correctly
    - Confirm that the `wc_address` contains the same address as the one stored in WooCommerce settings
3. Update the phone number using the `POST /wc/gla/mc/contact-information` endpoint
   - Confirm that the phone number is updated using the above `GET` endpoint
   - Confirm that the phone number is updated in Merchant Center : 
     <https://merchants.google.com/mc/merchantprofile/businessinfo>
   - Confirm that the address is updated in Merchant Center and it's the same as your WooCommerce address

### CODE REVIEW

**Comment by mikkamp** [comment]
The PR looks great although I haven't had the chance to test if fully. But how much work would it be to refactor it slightly to have `MerchantVerification::get_contact_information` return both addresses and the comparison? And have the controller just handle wrapping the final data in a REST response. I believe we built similar patterns for other controllers.

I'd like to use the contact information and address comparison internally to prompt a notification, and if it's only accessible through the controller then I have to wrap it in a REST request (and unwrap the response).

**Comment by nima-karimi** [comment]
@mikkamp I added a new `AddressUtility` class that contains the comparison functionality. The two addresses for Merchant Center and WooCommerce are obtainable from `MerchantVerification::get_contact_information` and `Settings::get_store_address` respectively.

File: src/API/Google/Settings.php [hunk 1]
```diff
@@ -349,6 +350,69 @@ protected function get_store_state(): string {
...
+	/**
+	 * Get the WooCommerce store physical address.
+	 *
+	 * @return AccountAddress
+	 *
+	 * @since x.x.x
+	 */
+	public function get_store_address(): AccountAddress {
+		/** @var WC $wc */
+		$wc = $this->container->get( WC::class );
```
Comment (mikkamp) [inline src/API/Google/Settings.php:362]: For other classes where we get a dependency from the container we add it in a [comment up top](https://github.com/woocommerce/google-listings-and-ads/blob/1.3.0/src/MerchantCenter/MerchantCenterService.php#L25-L32) so we don't have to look through the code to find them. Eventually we might want to refactor this class to either be ContainerAware or get the dependencies directly, but I guess we have #563 for that.

**Comment by eason9487** [comment]
https://github.com/woocommerce/google-listings-and-ads/blob/ad11cb1ebd127a708e5cf0b93e4f5bdba90d044b/src/MerchantCenter/MerchantVerification.php#L60

I `POST /wc/gla/mc/contact-information` with body `{ "phone_number": null }` but got "Invalid phone number." error. Probably need to adjust it because it should not update phone number to GMC if the user has already set a verified phone number via GMC.

**Review by mikkamp — COMMENTED** [review]
I based my PR off of this one so I could reuse some of the code. When testing I noticed a couple minor things.

I accidentally tried updating using `phone-number` instead of `phone_number` which gives me a fatal:
```
Argument 1 passed to Automattic\WooCommerce\GoogleListingsAndAds\MerchantCenter\MerchantVerification::update_contact_information() must be of the type string, null given
```
We might want to see if that needs to be a required parameter or if we can add another check so it doesn't fatal.

For the `ContactInformationControllerTest` it fails because it's not passed the fourth parameter in the constructor.

**Comment by nima-karimi** [comment]
If we make the phone number optional in the `POST` API by checking if it's not null then how would we actually set it to null? For example, if a user wants to remove its phone number?

File: src/API/Google/Settings.php (hunk 1 above)
Comment (nima-karimi) [reply src/API/Google/Settings.php:362]: Yeah, it should eventually be refactored that's why I didn't make any more changes than I had to. The `WC` class was already being used in this service so this isn't a new addition.

**Comment by nima-karimi** [comment]
Added some new changes to take into account null or empty phone numbers:

> Create separate methods to update phone number and address

> Only update the phone number if a value is set in the request params. If a null or empty value is set then remove the phone number from contact information.

> Rename the `MerchantVerification` service to `ContactInformation`.

**Comment by nima-karimi** [comment]
@eason9487

 If you provide a null or empty phone number this API will now set the phone number stored in MC to null. This can be used in case the user wants to remove their phone number. 

However, if you only want to update the address without updating the phone number, then do not provide a `phone_number` key in the JSON request body. If no `phone_number` key is present in the request, then it will remain unchanged.

@mikkamp I believe this commit (https://github.com/woocommerce/google-listings-and-ads/pull/919/commits/c558b2f13767d416612a8274c5ab1c420ab720d5) resolves the error you were experiencing too.

File: tests/Unit/API/Site/Controllers/MerchantCenter/ContactInformationControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	const ROUTE = '/wc/gla/mc/contact-information';
+
+	/**
+	 * Runs before each test is executed.
+	 */
+	public function setUp() {
+		parent::setUp();
+		$this->contact_information            = $this->createMock( ContactInformation::class );
+		$this->google_settings                = $this->createMock( Settings::class );
+		$this->address_utility                = $this->createMock( Settings::class );
```
Comment (mikkamp) [inline tests/Unit/API/Site/Controllers/MerchantCenter/ContactInformationControllerTest.php]: Just a small typo here:
```suggestion
		$this->address_utility                = $this->createMock( AddressUtility::class );
```

**Review by mikkamp — APPROVED** [review]
Looks good, I've tested it extensively with my PR based off this one and I don't find any other issues.

Just one small typo left to make the tests pass.

**Pull request merged**